File: Android/core/src/main/java/com/khushi/conduit/core/protocol/Names.kt
package com.khushi.conduit.core.protocol

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * The vocabulary shared by every Conduit interface. Spec:
 * Shared/Protocol/README.md §6 and §8. Names are open sets — a newer peer may
 * send one this build does not know — so they travel as strings.
 */
object ActionName {
    const val SESSION_PING = "session.ping"
    const val STATE_SYNC = "state.sync"
    const val MIRRORING_REQUEST = "mirroring.request"
    const val TRACKPAD_START = "trackpad.start"
    const val TRACKPAD_STOP = "trackpad.stop"
    const val LINK_SEND = "link.send"
    const val CLIPBOARD_SET = "clipboard.set"
    const val CALL_ANSWER = "call.answer"
    const val CALL_DECLINE = "call.decline"
    const val CALL_END = "call.end"
    const val CALL_DIAL = "call.dial"
    const val CALL_MUTE = "call.mute"
    const val MAC_FIND = "mac.find"
}

object EventName {
    const val SESSION_BYE = "session.bye"
    const val DEVICE_SNAPSHOT = "device.snapshot"
    const val DEVICE_BATTERY = "device.battery"
    const val DEVICE_NETWORK = "device.network"
    const val DEVICE_LOCK = "device.lock"
    const val FEATURES_CHANGED = "features.changed"
    const val TRACKPAD_STATE = "trackpad.state"
    const val TRACKPAD_CONFIG = "trackpad.config"
    const val CALL_INCOMING = "call.incoming"
    const val CALL_STATE = "call.state"
}

enum class FeatureId(val wire: String) {
    MIRRORING("mirroring"),
    REMOTE_INPUT("remoteInput"),
    TRACKPAD("trackpad"),
    CLIPBOARD("clipboard"),
    CAMERA("camera"),
    CALLS("calls"),
    LINKS("links"),
    AUDIO("audio"),
    FILES("files"),
    NOTIFICATIONS("notifications"),
    FIND_MAC("findMac");

    companion object {
        fun fromWire(value: String): FeatureId? = entries.firstOrNull { it.wire == value }
    }
}

@Serializable
enum class Availability(val wire: String) {
    @SerialName("available") AVAILABLE("available"),
    @SerialName("active") ACTIVE("active"),
    @SerialName("requires_permission") REQUIRES_PERMISSION("requires_permission"),
    @SerialName("requires_setup") REQUIRES_SETUP("requires_setup"),
    @SerialName("disabled") DISABLED("disabled"),
    @SerialName("unsupported") UNSUPPORTED("unsupported"),
    @SerialName("planned") PLANNED("planned");

    val isUsable: Boolean get() = this == AVAILABLE || this == ACTIVE
}

object ErrorCode {
    const val INVALID_REQUEST = "invalid_request"
    const val UNSUPPORTED = "unsupported"
    const val PERMISSION_DENIED = "permission_denied"
    const val REQUIRES_SETUP = "requires_setup"
    const val DEVICE_UNAVAILABLE = "device_unavailable"
    const val PHONE_LOCKED = "phone_locked"
    const val TIMEOUT = "timeout"
    const val NETWORK_CHANGED = "network_changed"
    const val PEER_RESTARTED = "peer_restarted"
    const val DUPLICATE_REQUEST = "duplicate_request"
    const val STALE_EVENT = "stale_event"
    const val NOT_PAIRED = "not_paired"
    const val VERSION_MISMATCH = "version_mismatch"
    const val BUSY = "busy"
    const val CANCELLED = "cancelled"
    const val INTERNAL = "internal"

    val known: Set<String> = setOf(
        INVALID_REQUEST, UNSUPPORTED, PERMISSION_DENIED, REQUIRES_SETUP, DEVICE_UNAVAILABLE,
        PHONE_LOCKED, TIMEOUT, NETWORK_CHANGED, PEER_RESTARTED, DUPLICATE_REQUEST, STALE_EVENT,
        NOT_PAIRED, VERSION_MISMATCH, BUSY, CANCELLED, INTERNAL,
    )
}

File: Shared/ConduitKit/Sources/ConduitProtocol/Names.swift
//
//  Names.swift
//  ConduitProtocol
//
//  Action, event, feature and error identifiers — the vocabulary shared by
//  every Conduit interface. Spec: Shared/Protocol/README.md §6 and §8.
//
//  Names are open sets: a newer peer may send one this build does not know,
//  and it must round-trip intact rather than fail to decode.
//

import Foundation

// MARK: - Actions

public struct ActionName: RawRepresentable, Hashable, Sendable, CustomStringConvertible {
    public let rawValue: String
    public init(rawValue: String) { self.rawValue = rawValue }
    public init(_ rawValue: String) { self.rawValue = rawValue }
    public var description: String { rawValue }

    public static let sessionPing = ActionName("session.ping")
    public static let stateSync = ActionName("state.sync")
    public static let mirroringRequest = ActionName("mirroring.request")
    public static let trackpadStart = ActionName("trackpad.start")
    public static let trackpadStop = ActionName("trackpad.stop")
    public static let linkSend = ActionName("link.send")
    public static let clipboardSet = ActionName("clipboard.set")
    public static let callAnswer = ActionName("call.answer")
    public static let callDecline = ActionName("call.decline")
    public static let callEnd = ActionName("call.end")
    public static let callDial = ActionName("call.dial")
    public static let callMute = ActionName("call.mute")
    public static let macFind = ActionName("mac.find")
}

// MARK: - Events

public struct EventName: RawRepresentable, Hashable, Sendable, CustomStringConvertible {
    public let rawValue: String
    public init(rawValue: String) { self.rawValue = rawValue }
    public init(_ rawValue: String) { self.rawValue = rawValue }
    public var description: String { rawValue }

    public static let sessionBye = EventName("session.bye")
    public static let deviceSnapshot = EventName("device.snapshot")
    public static let deviceBattery = EventName("device.battery")
    public static let deviceNetwork = EventName("device.network")
    public static let deviceLock = EventName("device.lock")
    public static let featuresChanged = EventName("features.changed")
    public static let trackpadState = EventName("trackpad.state")
    public static let trackpadConfig = EventName("trackpad.config")
    public static let callIncoming = EventName("call.incoming")
    public static let callState = EventName("call.state")
}

// MARK: - Features

public enum FeatureID: String, CaseIterable, Codable, Sendable {
    case mirroring
    case remoteInput
    case trackpad
    case clipboard
    case camera
    case calls
    case links
    case audio
    case files
    case notifications
    case findMac
}

public enum Availability: String, Codable, Sendable {
    case available
    case active
    case requiresPermission = "requires_permission"
    case requiresSetup = "requires_setup"
    case disabled
    case unsupported
    case planned

    /// True when the user can use the feature right now.
    public var isUsable: Bool { self == .available || self == .active }
}

// MARK: - Errors

public struct ErrorCode: RawRepresentable, Hashable, Codable, Sendable, CustomStringConvertible {
    public let rawValue: String
    public init(rawValue: String) { self.rawValue = rawValue }
    public init(_ rawValue: String) { self.rawValue = rawValue }
    public var description: String { rawValue }

    public init(from decoder: Decoder) throws {
        rawValue = try decoder.singleValueContainer().decode(String.self)
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        try c.encode(rawValue)
    }

    public static let invalidRequest = ErrorCode("invalid_request")
    public static let unsupported = ErrorCode("unsupported")
    public static let permissionDenied = ErrorCode("permission_denied")
    public static let requiresSetup = ErrorCode("requires_setup")
    public static let deviceUnavailable = ErrorCode("device_unavailable")
    public static let phoneLocked = ErrorCode("phone_locked")
    public static let timeout = ErrorCode("timeout")
    public static let networkChanged = ErrorCode("network_changed")
    public static let peerRestarted = ErrorCode("peer_restarted")
    public static let duplicateRequest = ErrorCode("duplicate_request")
    public static let staleEvent = ErrorCode("stale_event")
    public static let notPaired = ErrorCode("not_paired")
    public static let versionMismatch = ErrorCode("version_mismatch")
    public static let busy = ErrorCode("busy")
    public static let cancelled = ErrorCode("cancelled")
    public static let `internal` = ErrorCode("internal")

    static let known: Set<ErrorCode> = [
        .invalidRequest, .unsupported, .permissionDenied, .requiresSetup, .deviceUnavailable,
        .phoneLocked, .timeout, .networkChanged, .peerRestarted, .duplicateRequest, .staleEvent,
        .notPaired, .versionMismatch, .busy, .cancelled, .internal,
    ]

    /// Spec §8: a code this build does not know is handled as `internal`.
    public var normalized: ErrorCode { Self.known.contains(self) ? self : .internal }
}

public struct ProtocolError: Error, Codable, Sendable, Equatable {
    public var code: ErrorCode
    /// Human-readable and safe to show to the user.
    public var message: String

    public init(_ code: ErrorCode, _ message: String) {
        self.code = code
        self.message = message
    }
}
